### ISSUE TITLE
[2.0.0] Feature detection for Adapters

### ISSUE THREAD

**Issue by AlCalzone** [ioBroker/ioBroker.js-controller#244]
Sometimes we introduce new functionality into the js Controller that adapters can rely on or remove old functionality. It would be helpful to have an easy way from inside adapters to check if the currently installed controller supports feature xyz. Eg an object with a set of Flags that is updated whenever major functionality is added that adapters may choose to rely on. 

Examples:
* Async/promisified versions of Adapter methods
* The upcoming centralized non install methods

**Comment by Apollon77** [comment]
@GermanBluefox  What do you think?

On one hand it makes sense, but do an adapter developer really develop different code when a feature is not there? Or woul/should an adapter developer not simply tell the needed js-controller version as dependency?

**Comment by AlCalzone** [comment]
I'm thinking in the sense of progressive enhancement, as a JS-controller updates are a bigger thing than adapter updates:
1. A big, shiny new feature is being added to js-controller, but it might take a while until it becomes "stable"
2. Dev sees it and wants to use it in his adapter
3. In the adapter, he tests if the feature is available. If yes, he uses it, if not he may choose how to proceed: less functionality or implement a "polyfill" (the polyfill is what all my adapters currently do)
4. The adapter gets used (meanwhile the adapter and the new feature is already being tested without requiring major system updates - some users might already use the new js-controller)
5. The js-controller with the new feature gets moved to stable
6. In his next update, the dev declares the new js-controller version as a requirement and removes the test

**Comment by GermanBluefox** [comment]
I must say, I developed only once the code for old and new js-controller. 
I don't think if this really will be used.

**Comment by AlCalzone** [comment]
@GermanBluefox if *you* want new functionality, you just add it to js-controller yourself and be done with it. I get that. I'm thinking more from the view of an external (adapter) developer.

That being said, I don't think this is high priority, but still nice to have.

**Comment by Apollon77** [comment]
I also had a thought here. If someone want to check methods he simply checks if the method is there. So i currently only see it interesting for two things:
1.) changed behavior of Core stuff where an adapter needs to react on. But this would mean we have broke stuff and normally we do not want that
2.) Derry if alias feature is existing ;-) (because it is a under the hood feature that do not expose something). This could be interesting  eg von the „Former“ Adapters like linkeddevices. They could behave as before on old controllers or use alias on newer ones. 

So my proposal would be a adapter.js method adapter.adapterSupports(„FEATURE_NAME“) which returns true/false

Alternative Name (more generic) Adapter.systemSupports or only adapter.supportsFeature

**Comment by AlCalzone** [comment]
> if (adapter.supportsFeature && adapter.supportsFeature(„ALIAS“)) ...

we could even go full circle :rofl: 
`if (adapter.supportsFeature && adapter.supportsFeature(„SUPPORTS_FEATURE“) && ...)`

**Comment by Apollon77** [comment]
lol ... but hey needed because old js-controller will not have the "adapter.supportsFeature" :-(

**Comment by Apollon77** [comment]
So @GermanBluefox @AlCalzone I'm not that decision willing today ... which name you like to see then I can build it into 2.0.0

**Comment by AlCalzone** [comment]
`supportsFeature` as a method name is fine. As I wrote on telegram, we need to give the relevant features like alias a name/ID everybody understands and add documentation (feature name/id, short feature description, which version was it added?) in a location where developers can easily find it.

**Comment by Apollon77** [comment]
Location = Readme of js-controller for now, later the dev part of the webpage (as soon as someone finds time to write it ;-)
Then I add "ALIAS" as first id there for now
